src/System.CommandLine/Invocation/MiddlewareOrder.cs

```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.CommandLine.Invocation
+{
+    public enum MiddlewareOrder : byte
+    {
+        ProcessStart = byte.MinValue,
```

Comment:
This effectively prevents anyone from injecting anything before us. Is there a reason for this restriction?

I would also suggest that MiddlewareOrder.Default should equal default(MiddlewareOrder).

Reply:
The default should be after most of the middleware we've defined, not at the beginning. The existing middleware that uses the `ProcessStart` and `ExceptionHandler` positions will be incorrect if other middleware executes before them. Similarly, if `ErrorReporting` happens before custom middleware, then the custom middleware won't be reached.

We're opening this up for advanced scenarios so with that in mind it's probably reasonable for `ProcessStart` to be at `MinValue + 10`.

Also, all of these could use better names.

Reply:
I agree that ProcesStart should be greater than the minimum possible value to allow for people to inject middleware before the default stuff (this is also making the assumption that they include the default middleware at all). I think it is important to not assume that our middleware will always be the first one that needs to be executed.

On the note of individual values, I do like the idea of spreading out the predefined values a bit. I would probably suggest giving an enum member to each of the built-int middleware that we have define. I think then rather than trying to reason able which middleware ends up at each value, you simply can look at the enum name and value and reason through it.

Reply:
Because the built-in middleware will certainly change over time, this would make breaking changes to this enum common. And for most cases the specific order isn't going to matter and should not be a contract.

Reply:
Because the built-in middleware will certainly change over time, this would make breaking changes to this enum common. And for most cases the specific order isn't going to matter and should not be a contract.